Core/src/IfcGeometryConverter/FaceConverter.h

```diff
@@ -768,132 +767,165 @@ namespace OpenInfraPlatform {
+					// if simple case: one triangle without inner bound (= hole)
+					//  -> simple direct triangle construction is possible (probably fast)
+					if ((faceBoundLoops.size() == 1) && (faceBoundLoops[0].size() == 3))
+					{
+						// std::vector<carve::geom::vector<3>> loopVertices3D = faceBoundLoops[0];
+						addTriangleToPolyhedronData(faceBoundLoops[0], polygon, polygonIndices);
+					}
+					else
+					{
+						// general case: arbitrary number of vertices, possible inner bound (= hole)
+						//  -> elaborate triangulation with respect to arbitrary number of vertices and holes is necessary;
 
-					// Save polygon indices of merged vertices
-					std::map<uint32_t, uint32_t> mergedIndices;
-					std::vector<EXPRESSReference<typename IfcEntityTypesT::IfcFaceBound>> modBounds;
-					modBounds.reserve(2);
-					bool faceLoopReversed = false;
+						addArbitraryFaceToPolyhedronData(face, faceBoundLoops, polygon, polygonIndices);
+					}
+				}
+				
+				void convertIfcFaceBoundList(
+					std::vector<EXPRESSReference<typename IfcEntityTypesT::IfcFaceBound>> ifcFaceBounds,
+					const carve::math::Matrix& pos,
+					std::vector<std::vector<carve::geom::vector<3>>>& faceBoundLoops) const noexcept(true)
+				{
+					// As carve expects outer boundary of face to be at first index, outer boundary has to be moved to index 0 and inner boundary has index >= 1
+					for (auto it = ifcFaceBounds.begin(); it != ifcFaceBounds.end(); it++)
+					{
+						if (it->isOfType<typename IfcEntityTypesT::IfcFaceOuterBound>())
+						{
+							// swap element 'it' with first element
+							std::rotate(ifcFaceBounds.begin(), it, it + 1);
 
-					// Loop through all boundary definitions
-					int boundID = -1;
+							// according to ifc-documentation, only one boundary (= loop) can be outer boundary:
+							// "One loop is optionally distinguished as the outer loop of the face."
+							// (https://standards.buildingsmart.org/IFC/DEV/IFC4_3/RC3/HTML/link/ifcface.htm)
+							// thus, break the for-loop after THE outer boundary was found
+							break;
+						}
+					}
 
-					// If polygon has more than 3 vertices, then we have to project polygon into 2D, so that carve can triangulate the mesh
-					ProjectionPlane plane = UNDEFINED;
+					faceBoundLoops.resize(ifcFaceBounds.size());
+					for (size_t i = 0; i < ifcFaceBounds.size(); i++)
+					{
+						// get vertices of one bound loop into faceBoundLoops[i]
+						convertIfcFaceBound(ifcFaceBounds[i], pos, faceBoundLoops[i]);
+					}
+				}
+
+				void convertIfcFaceBound(
+					const EXPRESSReference<typename IfcEntityTypesT::IfcFaceBound>& bound,
+					const carve::math::Matrix& pos,
+					std::vector<carve::geom::vector<3>>& loopVertices3D) const noexcept(false)
+				{
+					//	IfcLoop   has subtypes   IfcEdgeLoop, IfcPolyLoop, IfcVertexLoop
+					const EXPRESSReference<typename IfcEntityTypesT::IfcLoop>& loop = bound->Bound;
+					bool polyOrientation = bound->Orientation;
+
+					if (!loop) {

```

Comment:
test, if reference expired

Reply:
Done in commit 785881d.
I added the expired-test of the `EXPRESSReference<... IfcFaceBound> bound` because it wasn't done before.
As well, I adjusted and kept the expired-test of the `EXPRESSReference<... IfcLoop> loop`, which is done immediately after it's definition. At this point, I'm not sure about it - I would expect this test at the beginning of the function `CurveConverter::convertIfcLoop`, however I can't spot it there.

Reply:
if the test is not at the beginning of the function, please put it there

Reply:
Done in commit 5d90a20.